packages/apps/sendgrid/src/functios-to-sendgrid/abandoned-carts.ts

```diff
@@ -0,0 +1,88 @@
+import api from '@cloudcommerce/api';
+// eslint-disable-next-line import/no-unresolved
+import { getFirestore } from 'firebase-admin/firestore';
+import { sgSendMail } from './utils';
+import parseCartToSend from './parse-data-to-send';
+
+export default async () => {
+  try {
+    logger.log('# Check abandoned carts');
+    const [application] = (await api.get(
+      'applications?app_id=111223&fields=hidden_data',
+    )).data.result;
+
+    const appData = {
+      ...application.data,
+      ...application.hidden_data,
+    };
+
+    let minCartDate = new Date();
+    minCartDate.setHours(minCartDate.getHours() - (appData.abandoned_cart_delay || 3));
+    const lastDateSent = (await getFirestore().doc('sendGrid/lastCartSent').get()).data();
+    if (lastDateSent && lastDateSent.created_at) {
+      const lastNotifiedCartDate = new Date(lastDateSent.created_at);
+      if (lastNotifiedCartDate.getTime() > minCartDate.getTime()) {
+        minCartDate = lastNotifiedCartDate;
+      }
+    }
+
+    const [
+      { data: { result: abandonedCarts } },
+      { data: store },
+    ] = await Promise.all([
+      api.get(`carts?completed=false&available=true&metafields.field!=sendgrid_sent&created_at>=${minCartDate.toISOString()}`),
+      api.get('stores/me'),
+    ]);
+
+    if (abandonedCarts.length && store) {
+      for (let i = 0; i < abandonedCarts.length; i++) {
+        const cart = abandonedCarts[i];
+        const customerId = cart.customers ? cart.customers[0] : null;
+
+        logger.log(`>> Preparing to send email to: ${customerId}, about cart:${cart._id} `);
+
+        const apiKey = appData.sendgrid_api_key;
+        const merchantEmail = appData.sendgrid_mail;
+        if (apiKey && merchantEmail && customerId) {
+          // eslint-disable-next-line no-await-in-loop
+          const customer = (await api.get(`customers/${customerId}`)).data;
+          if (cart && customer) {
+            const emailData = parseCartToSend(appData, 'abandoned_cart', cart, store, customer);
+            if (emailData) {
+              sgSendMail(emailData, apiKey)
+                .then(() => {
+                  logger.log('>> Email sent ');
+
+                  api.post(
+                    `carts/${cart._id}/metafields`,
+                    {
+                      field: 'sendgrid_sent',
+                      value: true,
+                    } as any,
+                  );
+
+                  if (i === (abandonedCarts.length - 1)) {
+                    // save last cart
+                    getFirestore()
+                      .doc('sendGrid/lastCartSent')
+                      .set({
+                        cartId: cart._id,
+                        create_at: cart.created_at,
+                      }).catch(logger.error);
+                  }
+                })
+                .catch((err) => {
+                  logger.error('>>#Error send email => ', err);
+                });
```

Comment:
Isso aqui tem potencial pra dar vários problemas 🙃 
Você tá lendo os carrinhos na API sem paginar e sem limitar, o que significa que você pode (e em lojas maiores vai) pegar muitos carrinhos na mesma request.
Em sequência você tenta disparar em concorrência (sem await) e-mails pra todos os carrinhos, o que provavelmente vai te gerar problemas com rate limit do SendGrid, seria o menor dos problemas porque se o SendGrid tiver um rate limit muito permissivo você em seguida tenta salvar o metafield nos carrinhos também em concorrência, dessa forma você vai ter problemas com rate limit na Store API não só na execução dessa função, mas como ela compartilha a autenticação com outros apps e até o SSR, essa execução tem potencial pra fazer inclusive a loja parar de funcionar 😄

Reply:
Nesse segundo comentário https://github.com/ecomplus/cloud-commerce/pull/62#discussion_r1001237860 o _parece melhor_ era justamente porque dessa outra forma reduzimos requests na API, algum motivo pra não escolher não fazer dessa forma?
Vi que você está salvando o doc no Firestore com o último carrinho processado, mas você não usa, se não for usar melhor não salvar na verdade, aí evita writes desnecessários no Firestore, acho que não faz sentido salvar o metafield e o `sendGrid/lastCartSent` no Firestore...

Reply:
De toda forma, principalmente se for salvar o metafield você precisa limitar os carrinhos listados, e deve ter que processar os emails um a um com `await` pra evitar problemas também com a API do SendGrid.

Reply:
> Nesse segundo comentário https://github.com/ecomplus/cloud-commerce/pull/62#discussion_r1001237860 o parece melhor era justamente porque dessa outra forma reduzimos requests na API, algum motivo pra não escolher não fazer dessa forma?

Se a nossa função está programada para executar 3 vezes ao dia e o lojista configurar para enviar email para carrinhos abandondos depois de 3 horas, isso não poder ou fazer envio de mais de um email para o mesmo carrinho ou até mesmo deixar de enviar para algum carrinho ?

Reply:
> isso não poder ou fazer envio de mais de um email para o mesmo carrinho ou até mesmo deixar de enviar para algum carrinho ?

Em qual cenário repetiria o envio? Poderia deixar de enviar se forem carrinhos demais e durante a execução não conseguir disparar todos, com o tempo iria acumulando, mas nesse caso em qualquer implantação sugerida aqui não resolveria na verdade.